Branch-SDK/src/io/branch/referral/SystemObserver.java

```diff
@@ -51,26 +57,86 @@ public boolean hasRealHardwareId() {
 	}
 	
 	public String getURIScheme() {
+	    return getURIScheme(context_.getPackageName());
+	}
+	
+	public String getURIScheme(String packageName) {
+		String scheme = BLANK;
 		PackageManager pm = context_.getPackageManager();
-	    try {
-	        ApplicationInfo ai = pm.getApplicationInfo(context_.getPackageName(), 0);
+		try {
+	        ApplicationInfo ai = pm.getApplicationInfo(packageName, 0);
 	        String sourceApk = ai.publicSourceDir;
+	        JarFile jf = null;
+	        InputStream is = null;
+	        byte[] xml = null;
 	        try {
-	            JarFile jf = new JarFile(sourceApk);
-	            InputStream is = jf.getInputStream(jf.getEntry("AndroidManifest.xml"));
-	            byte[] xml = new byte[is.available()];
+	            jf = new JarFile(sourceApk);
+	            is = jf.getInputStream(jf.getEntry("AndroidManifest.xml"));
+	            xml = new byte[is.available()];
                 //noinspection ResultOfMethodCallIgnored
                 is.read(xml);
-	            String scheme = new ApkParser().decompressXML(xml);
-	            jf.close();
-	            is.close();
-	            return scheme;
-	          } catch (Exception ignored ) {
-	        	  
-	          }
-	    } catch (NameNotFoundException ignored ) {
+	            scheme = new ApkParser().decompressXML(xml);
+	        } catch (Exception ignored) {
+	        } finally {
+	        	xml = null;
+	        	try {
+	        		jf.close();
+	        		is.close();
```

Comment:
do we want to set jf and is to null as well to ensure there aren't any dangling pointers?